0000-friends_in_high_kindednesses.md

```diff
@@ -0,0 +1,442 @@
+- Feature Name: associated_type_constructors
+- Start Date: 2016-04-29
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow type constructors to be associated with traits. This is an incremental
+step toward a more general feature commonly called "higher-kinded types," which
+is often ranked highly as a requested feature by Rust users. This specific
+feature (associated type constructors) resolves one of the most common use
+cases for higher-kindedness, is a relatively simple extension to the type
+system compared to other forms of higher-kinded polymorphism, and is forward
+compatible with more complex forms of higher-kinded polymorphism that may be
+introduced in the future.
+
+
+# Motivation
+[motivation]: #motivation
+
+Consider the following trait as a representative motivating example:
+
+```rust
+trait StreamingIterator {
+    type Item<'a>;
+    fn next<'a>(&'a mut self) -> Option<Self::Item<'a>>;
```

Comment:
> Can you explain this comment? That is the fundamental concept of this feature.

For one thing, I expect the compiler warn about `type Foo<'a>` since `'a` is introduced and named but not used. (Note that the compiler currently doesn't complain about unused lifetimes or type parameters in many cases that it should, IMO.) `type Foo<'_>` makes it clearer that the lifetime is used as a placeholder.

I have typically read `type Foo<'a> = &'a [u8];` and similar things as introducing the type parameter on the left hand side and then using it on the right hand side, but in `type Foo<'a>;` the `'a` isn't an introduction to any later use. (Sorry this probably isn't super clear.)